src/ellswift.rs

```diff
@@ -0,0 +1,705 @@
+//!   a square root of -3.
+//! - If u=0, set u=1 instead.
+//! - If t=0, set t=1 instead.
+//! - If u^3 + t^2 + 7 = 0, multiply t by 2.
+//! - Let X = (u^3 + 7 - t^2) / (2 * t)
+//! - Let Y = (X + t) / (C * u)
+//! - Return the first in [u + 4 * Y^2, (-X/Y - u) / 2, (X/Y - u) / 2] that is an
+//!   X coordinate on the curve (at least one of them is, for any u and t).
+//!
+//! Then an `ElligatorSwift` encoding of x consists of the 32-byte big-endian
+//! encodings of field elements u and t concatenated, where f(u,t) = x.
+//! The encoding algorithm is described in the paper, and effectively picks a
+//! uniformly random pair (u,t) among those which encode x.
+//!
+//! If the Y coordinate is relevant, it is given the same parity as t.
+//!
+//! Changes w.r.t. the the paper:
+//! - The u=0, t=0, and u^3+t^2+7=0 conditions result in decoding to the point
+//!   at infinity in the paper. Here they are remapped to finite points.
+//! - The paper uses an additional encoding bit for the parity of y. Here the
+//!   parity of t is used (negating t does not affect the decoded x coordinate,
+//!   so this is possible).
+
+use core::fmt::{self, Display, Formatter};
+use core::ptr;
+use core::str::FromStr;
+
+use ffi::CPtr;
+use secp256k1_sys::types::{c_int, c_uchar, c_void};
+
+use crate::{constants, ffi, from_hex, Context, Error, PublicKey, Secp256k1, SecretKey};
+
+unsafe extern "C" fn hash_callback<F>(
+    output: *mut c_uchar,
+    x32: *const c_uchar,
+    ell_a64: *const c_uchar,
+    ell_b64: *const c_uchar,
+    hash_func: *mut c_void,
+) -> c_int
+where
+    F: FnMut([u8; 32], [u8; 64], [u8; 64]) -> ElligatorSwiftSharedSecret,
+{
+    let callback: &mut F = &mut *(hash_func as *mut F);
+    let mut x32_array = [0u8; 32];
+    let mut ell_a64_array = [0u8; 64];
+    let mut ell_b64_array = [0u8; 64];
+
+    // Copy the data into Rust slices
+    ptr::copy_nonoverlapping(x32, x32_array.as_mut_c_ptr(), 32);
+    ptr::copy_nonoverlapping(ell_a64, ell_a64_array.as_mut_c_ptr(), 64);
+    ptr::copy_nonoverlapping(ell_b64, ell_b64_array.as_mut_c_ptr(), 64);
+    // Call the hash function that was passed in through the `data` pointer
+    let secret = callback(x32_array, ell_a64_array, ell_b64_array);
+    // Copy the output from a [ElligatorSwiftSharedSecret] into the output pointer
+    ptr::copy_nonoverlapping(secret.0.as_ptr(), output, secret.0.len());
+
+    secret.0.len() as c_int
+}
+
+/// `ElligatorSwift` is an encoding of a uniformly chosen point on the curve
+/// as a 64-byte array that is indistinguishable from a uniformly random.
+/// This object holds two field elements u and t, which are the inputs to
+/// the `ElligatorSwift` encoding function.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ElligatorSwift(ffi::ElligatorSwift);
+
+impl ElligatorSwift {
+    /// Create a new `ElligatorSwift` object from a 64-byte array.
+    pub fn ellswift_create(priv32: SecretKey, rand32: [u8; 32]) -> ElligatorSwift {
```

Comment:
I think we should be using Rust names for all the methods on the `ElligatorSwift` type instead of mirroring the C names, e.g., this one could be `new`. Sorry, didn't catch that on last review. And perhaps we should use `sk` or `secret_key` for the `SecretKey` parameter name. (`rand32` reads a bit like C too :)